tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -422,29 +595,131 @@ void RunSwapDimension1And2InTensor3(const GPUDevice& d, const T* input,
   // If both dimensions are not trivial, use tiles for the actual swapping.
   // Otherwise, the trivial swapping relying on the ldg cache is more efficient.
   static const int kMinDimensionToUseTiles = 16;
-  bool use_tiles = (input_dims[1] >= kMinDimensionToUseTiles &&
-                    input_dims[2] >= kMinDimensionToUseTiles);
-  if (use_tiles) {
+  static const int kMinDimensionToUseRectTiles = 96;
+
+  bool large_matrix = (input_dims[1] >= kMinDimensionToUseTiles &&
+                       input_dims[2] >= kMinDimensionToUseTiles);
+  bool narrow_matrix = (input_dims[1] >= kMinDimensionToUseRectTiles ||
+                      input_dims[2] >= kMinDimensionToUseRectTiles);
+  if (large_matrix) {
     // We get best performance when TileSize is the number of threads in a warp
     // (32 on our GPUs) and NumSubTiles is 8, so our block size is 8 * 32 = 256
     // threads.
     static const int TileSize = 32;
-    static const int NumSubTiles = 8;
+    static const int ThreadNum = 256;
```

Comment:
Hm, `ThreadNum` sounds to me like "this thread's number" rather than "number of threads".  But any reason not to call this `BlockSize`?

Reply:
Because in general the imaginary 2-dimensional thread blocks can change shape for reading and outputting matrices and therefore do not have a fixed size (essentially the effective_thread_num is the size of the imaginary block size, which can be different in the same kernel execution depending on whether you are reading or writing). I'm afraid that calling it BlockSize may induce some incorrect intuition.